animate/checkpointing.py

```diff
@@ -0,0 +1,97 @@
+from .metric import RiemannianMetric
+from .utility import get_checkpoint_dir
+import firedrake
+import firedrake.checkpointing as fchk
+import firedrake.function as ffunc
+import os
+
+__all__ = ["load_checkpoint", "save_checkpoint"]
+
+
+def _fix_checkpoint_filename(filename):
+    """
+    Convert a checkpoint filename to absolute form.
+
+    :arg filename: the filename without its path
+    :type filename: :class:`str`
+    :returns: the absolute filename
+    :rtype: :class:`str`
+    """
+    if "/" in filename:
+        raise ValueError(
+            "Provide a filename, not a filepath. Checkpoints will be stored in"
+            f" '{get_checkpoint_dir()}'."
+        )
+    name, ext = os.path.splitext(filename)
+    ext = ext or ".h5"
+    if ext != ".h5":
+        raise ValueError(f"File extension '{ext}' not recognised. Use '.h5'.")
+    return os.path.join(get_checkpoint_dir(), name + ext)
+
+
+def load_checkpoint(filename, metric_name, comm=firedrake.COMM_WORLD):
+    """
+    Load a metric from a :class:`~.CheckpointFile`.
+
+    Note that the checkpoint will have to be stored within Animate's ``.checkpoints``
+    subdirectory.
+
+    :arg filename: the filename of the checkpoint
+    :type filename: :class:`str`
+    :arg metric_name: the name used to store the metric
+    :type metric_name: :class:`str`
+    :kwarg comm: MPI communicator for handling the checkpoint file
+    :type comm: :class:`mpi4py.MPI.Intracom`
+    :returns: the metric loaded from the checkpoint
+    :rtype: :class:`animate.metric.RiemannianMetric`
+    """
+    fname = _fix_checkpoint_filename(filename)
+    if not os.path.exists(fname):
+        raise Exception(f"Metric file does not exist! Path: {fname}.")
+    with fchk.CheckpointFile(fname, "r", comm=comm) as chk:
+        mesh = chk.load_mesh()
```

Comment:
--- Marking all connected comments with these dashes---

L32 should be `def load_checkpoint(filename, mesh_name, metric_name, comm=firedrake.COMM_WORLD):`

and L52 `mesh = chk.load_mesh(mesh_name)`

Reply:
I couldn't get this to work. I won't have chance to work on this for a few days so if you have a working solution please feel free to push to this branch.

Reply:
Fixed in 7fcce996cd69fb074e6bf6657c6346631053461f .

Here is the MFE if you want to compare between commits:
```
old_mesh = UnitSquareMesh(3, 3, name="old_mesh")
metric = RiemannianMetric(TensorFunctionSpace(old_mesh, "CG", 1))
metric.set_parameters({})
new_mesh = adapt(old_mesh, metric, name="new_mesh")
```
Gives:
```
Traceback (most recent call last):
  File "/data2/testing/par/animate/mfe.py", line 9, in <module>
    new_mesh = adapt(old_mesh, metric, name="new_mesh")
  File "/home/ubuntu/software/firedrake-feb24/src/animate/animate/adapt.py", line 203, in adapt
    metric0 = load_checkpoint(metric_fname, metric_name, comm=COMM_SELF)
  File "/home/ubuntu/software/firedrake-feb24/src/animate/animate/checkpointing.py", line 52, in load_checkpoint
    mesh = chk.load_mesh()
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/ubuntu/software/firedrake-feb24/src/firedrake/firedrake/checkpointing.py", line 1035, in load_mesh
    tmesh_name = self._get_mesh_name_topology_name_map()[name]
KeyError: 'firedrake_default'
```
because it expects that `old_mesh.name = 'firedrake_default'`.

Reply:
@jwallwork23 if you think this is confusing for future users/developers I can also make some comments and give more descriptive variable names. Happy to do it if you want me to :)